pkg/build/controller/build/build_controller.go

```diff
@@ -372,14 +383,24 @@ func (bc *BuildController) createPodSpec(originalBuild *buildapi.Build, ref stri
 	return podSpec, nil
 }
 
-// resolveOutputDockerImageReference returns a reference to a Docker image
-// computed from the buid.Spec.Output.To reference.
-func (bc *BuildController) resolveOutputDockerImageReference(build *buildapi.Build) (string, error) {
+// resolveOutputDockerImageReference returns a Docker image name
+// computed from the build.Spec.Output.To reference.  It also attempts to find a Secret
+// associated with the build service account which can push to the image location and
+// return that, if found.
+// Note that we are using controller level permissions to resolve the imagestream and secret,
+// meaning users could theoretically define a build that references an imagestream they cannot
+// see, and 1) get the docker image reference of that imagestream and 2) a reference to a secret
+// associated with a service account that can push to that location.  However they still cannot view the secret,
+// and ability to use a service account implies access to its secrets, so this is considered safe.
+// Furthermore it's necessary to enable triggered builds since a triggered build is not "requested"
+// by a particular user, so there are no user permissions to validate against in that case.
+func (bc *BuildController) resolveOutputDockerImageReference(build *buildapi.Build) (string, *kapi.LocalObjectReference, error) {
 	outputTo := build.Spec.Output.To
 	if outputTo == nil || outputTo.Name == "" {
-		return "", nil
+		return "", build.Spec.Output.PushSecret, nil
```

Comment:
I don't think it should be.  If for whatever reason the user did specify a spec.Output.PushSecret (but no Spec.Output.To) we should reflect that in their build.Spec.  The consumer of this function is in part responsible for setting the build.Spec.Output.PushSecret based on the buildconfig.Spec.Output.PushSecret, so if we returned nil here, build.Spec.Output.PushSecret would not get set properly (again, there's no current purpose to the value, but not setting it would be strange since the user put it in their BC)


imo the contract of this function is:
1) figure out the image push target
2) return the secret to use, defaulting to the secret provided by the user in the build if any

(it is true that the code that uses this function will, on seeing a "nil" pushSecret, decline to update the build, so strictly speaking returning nil here is not a problem, but I still prefer the behavior of the function to be "default to the user provided secret" in all cases, not just some cases)